tests/pgshovel/replication/streams/kafka.py

```diff
@@ -106,9 +118,37 @@ def test_crashes_on_no_state(stream):
         next(stream.consume(None))
 
 
-def test_crashes_on_bad_state(writer, stream):
-    writer.push(transaction)
-    state = State()
-    state.stream_state.transaction_state.in_transaction.publisher = 'junk'
-    with pytest.raises(InvalidPublisher):
-        next(stream.consume(state))
\ No newline at end of file
+def test_validates_stream_and_crashes_when_invalid(writer, stream, state):
+    messages = list(islice(transactions(), 3))
+    messages[1] = messages[0]
+
+    writer.push(messages)
+
+    with pytest.raises(SequencingError):
+        list(stream.consume(state))
+
+
+def test_discards_messages_until_start_of_transaction(writer, stream, state, sliced_transaction):
+    writer.push(sliced_transaction)
+
+    consumed = list(islice(stream.consume(state), 3))
+    assert [message for _, _, message in consumed] == sliced_transaction[-3:]
+
+
+def test_discarded_messages_is_configurable(configuration, cluster, client, state, writer, sliced_transaction):
+    writer.push(sliced_transaction)
+
+    configuration['prime_threshold'] = 1
+    ansy_stream = KafkaStream.configure(configuration, cluster, 'default')
+
+    less_ansy_config = configuration.copy()
+    less_ansy_config['prime_threshold'] = 2
+    less_ansy_stream = KafkaStream.configure(less_ansy_config, cluster, 'default')
+
+    client.ensure_topic_exists(ansy_stream.topic)
+
+    with pytest.raises(InvalidEventError):
```

Comment:
This relies on the validation error to catch the lack of `BeginOperation` occurring. Wondering if it would be better for the priming function to raise an error if it couldn't prime successfully?